### PR TITLE
Docker support for multi stage builds for Dev and Prod environments

### LINKED ISSUE #1751: Implementing multi stage builds to Dockerfile
**Is your feature request related to a problem? Please describe.**
Currently, our Dockerfile lacks the implementation of multi-stage builds, leading to larger image sizes. This not only affects the efficiency of our deployment process but also contributes to increased resource consumption on cloud instances. Large Docker images can negatively impact both performance and security in cloud environments.

**Describe the solution you'd like**
Implement multi-stage builds in the Dockerfile to optimize image size, improve deployment speed, and enhance the overall efficiency of our containerized applications. Multi-stage builds allow us to use separate stages in the Dockerfile, enabling us to build and compile code in one stage and then copy only the necessary artifacts to the final image. This will result in smaller image sizes without compromising the functionality of the application.

**Describe alternatives you've considered**
One alternative is to manually clean up unnecessary files and dependencies after building the application, but this can be error-prone and time-consuming. Multi-stage builds offer a more automated and standardized solution to achieve smaller and more optimized images.

**Approach to be followed (optional)**
1.   Identify the different stages required for building and running the application.
2.   Implement the necessary build and compile steps in the first stage.
3.   Copy only the essential artifacts from the first stage to the final image in the second stage.
4.   Ensure that the final image contains only the dependencies and files required for runtime.
5.   Test the multi-stage build locally to verify its effectiveness and correctness.
6.   Update the documentation to reflect the changes made to the Dockerfile.

**Additional context**
Implementing multi-stage builds is crucial for optimizing our Docker images, especially in cloud instances where image size directly impacts resource utilization and deployment speed. Smaller images not only contribute to faster deployment but also improve the security and performance of our cloud API instances. The reduction in image size leads to decreased network transfer times, lower storage costs, and a minimized attack surface, ultimately enhancing the overall efficiency of our containerized applications in cloud environments.

### PR DESCRIPTION
**Pull request by varshith257**
**What kind of change does this PR introduce?**

Implemented multi-stage builds for Dockerfile that reduces docker image size

**Issue Number:**

Fixes #1751 

**Did you add tests for your changes?**

Yes

**Snapshots/Videos:**

image

**Demo**

https://github.com/PalisadoesFoundation/talawa-api/assets/117595548/50234409-b41c-4f6a-8701-e5b92fe767dd

**If relevant, did you update the documentation?**

Will be updated soon after creating Dockerfile for the production environment

**Other information**

Next PR will be raised for Docker support to the Production environment as part of this issue

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**
Yes

### CODE REVIEW

**Comment by varshith257** [comment]
@palisadoes
> [!WARNING]
 Getting the following error when  running `npm run build`  
image

**Comment by varshith257** [comment]
> [!NOTE]
Resolved by adding `"resolveJsonModule": true` in the compiler options in tsconfig.build.json file

```
{
  "compilerOptions": {
    "resolveJsonModule": true,
  }
}

```

**Comment by kb-0311** [comment]
@varshith257 In the file `tsconfig.build.json` , add `    "resolveJsonModule": true` line in the `compilerOptions` object this will get rid of the error.

**Comment by kb-0311** [comment]
Are you planning on shifting the content of the previous `Dockerfile` to `Dockerfile.prod` in this PR itself?

**Comment by varshith257** [comment]
> Are you planning on shifting the content of the previous `Dockerfile` to `Dockerfile.prod` in this PR itself?


> Are you planning on shifting the content of the previous `Dockerfile` to `Dockerfile.prod` in this PR itself?

Planning to open a new PR for prod env to avoid conflicts

**Comment by varshith257** [comment]
> Wouldn't that break the current deployment pipeline for the time duration you are assigned to the issue?

I think no because Dockerfile.dev and Dockerfile.prod are interrelated

**Comment by varshith257** [comment]
@kb-0311  I have done setup with production environment. It says App server is online but unable to access it .


```
PS C:\Users\91970\Desktop\talawa\talawa-api> npm run build

> talawa-api@1.0.0 prebuild
> graphql-codegen && rimraf ./build

✔ Parse Configuration
✔ Parse Configuration
✔ Generate outputs

> talawa-api@1.0.0 build
> tsc --pretty --project tsconfig.build.json

PS C:\Users\91970\Desktop\talawa\talawa-api> npm run start

> talawa-api@1.0.0 start
> cross-env pm2-runtime start ./build/server.js --watch

2024-02-24T13:32:03: PM2 log: Launching in no daemon mode
2024-02-24T13:32:03: PM2 log: [Watch] Start watching server
2024-02-24T13:32:03: PM2 log: App [server:0] starting in -fork mode-
2024-02-24T13:32:03: PM2 log: App [server:0] online
```

**Comment by varshith257** [comment]
@palisadoes @noman2002 @kb-0311 @vasujain275 What is the context of the production environment in Talawa-API?

File: docker-compose.yaml
Comment (vasujain275) [inline docker-compose.yaml:21]: We can't specify the Dev Docker file in this docker-compose as it will affect our current cloud instance deployments

File: Dockerfile.dev
Comment (vasujain275) [inline Dockerfile.dev:20]: You can't have node:alpine as final image for “Development” Dockerfile as it will cause problems in dev environment

**Comment by Cioppolo14** [comment]
@SiddheshKukade @EshaanAgg Can you review this PR?

**Comment by vasujain275** [comment]
@varshith257 @kb-0311 

> > Wouldn't that break the current deployment pipeline for the time duration you are assigned to the issue?
> 
> I think no because Dockerfile.dev and Dockerfile.prod aren't interrelated

1. Yes, the Dockerfile.dev and Dockerfile.prod aren't interrelated. But the current deployment pipeline is dependent on docker-compose file you modified. 
2. This will break the current pipeline and directly modifying the docker compose file is not a good approach imo.

---

> @palisadoes @noman2002 @kb-0311 @vasujain275 What is the context of the production environment in Talawa-API?

You can get context for this here - #1428 

---

So, I think there is some confusion regarding the purpose of this pr. Here is what I understand from my point of view - 
1. This PR need to delete the current **Dockerfile** and add **two** new Dockerfiles - 
    - Dockerfile.dev
    - Dockerfile.prod
2. The `Dockerfile.dev` should have code mounted as volume and use node:lts as final base image.
4. The `Dockerfile.prod` should only have build files at the final stage and use node:alpine as final base image.
5. Once these both Dockerfiles are added, we can proceed with using Dockerfile.prod in the deployment CI/CD pipeline.
6. I think this approach can be good for that - https://github.com/PalisadoesFoundation/talawa-api/issues/1751#issuecomment-1961783875, but we can improve this approach by discussing it here at:  #1428

File: docker-compose.yaml
Comment (varshith257) [reply docker-compose.yaml:21]: If there is only one Dockerfile we can remove the specifying of dockerfile. But here we have two docker files hence it necessary to specify which dockerfile it should should to build image

Comment (vasujain275) [reply docker-compose.yaml:21]: Ye should have two docker-compose files too - `docker-compose.dev.yml` and `docker-compose.prod.yml`

Comment (varshith257) [reply docker-compose.yaml:21]: I am thinking about it .Ok let us use then two docker-compose files

Comment (varshith257) [reply docker-compose.yaml:21]: @vasujain275  I didn't modify much about docker-compose for it. Just introduced volumes and use of specific docker file to build image.

**Comment by varshith257** [comment]
@kb-0311 @vasujain275 I think done this with the development environment Docker Setup.
Let me make another PR after this merged to avoid further conflicts

**Comment by vasujain275** [comment]
@varshith257 I think its best to add the changes to this pr only. We will lose all previous conversations with new pr.

**Comment by varshith257** [comment]
> @varshith257 I think its best to add the changes to this pr only. We will lose all previous conversations with new pr.

Yes, But there are merge conflicts raised and when solving it made unnecessary deletions of files and I have to delete this branch once it merged if its ok with develop env Docker support

**Comment by vasujain275** [comment]
@varshith257 I think `Dockerfile.prod` and `docker-compose.prod.yaml` can be added to this pr only without any such conflicts, rather than a different pr as this pr is incomplete without them. But I might be wrong, Please discuss this with @kb-0311 and @noman2002

**Comment by kb-0311** [comment]
@varshith257 Yes I agree with @vasujain275  you can add the prod file in this PR itself, with no issues. if there are merge conflicts we can solve that.

**Comment by varshith257** [comment]
@kb-0311 @vasujain275 Done with all changes as we discussed

**Comment by kb-0311** [comment]
LGTM , @vasujain275 anything you want to add?

**Comment by palisadoes** [comment]
1. Is there a way to put the docker files in a separate folder so that the root directory becomes less cluttered?
2. This approach will also help our PR authorized file workflow because we can check the folder for changes versus individual files in the root directory

**Comment by vasujain275** [comment]
@palisadoes Yes we can move all the docker files to a directory, that will be a good approach.

@varshith257 Please move all the docker files to a directory named `docker` and change the path of dockerfiles in docker-compose accordingly. You need to change the path and context parameters I think.

**Comment by varshith257** [comment]
> 1. Is there a way to put the docker files in a separate folder so that the root directory becomes less cluttered?
> 2. This approach will also help our PR authorized file workflow because we can check the folder for changes versus individual files in the root directory

@palisadoes Moved all files into docker folder as suggested

File: docker/docker-compose.dev.yaml
Comment (vasujain275) [inline docker/docker-compose.dev.yaml:20]: Context shouldn't be `..` instead of `.` ?

File: docker/docker-compose.prod.yaml
Comment (vasujain275) [inline docker/docker-compose.prod.yaml:19]: Context shouldn't be `..` instead of `.` ?

**Comment by varshith257** [comment]
@kb-0311 @vasujain275 @noman2002 A thing to get clarified, Are we going to release production build of Talawa-Api by setup of VPS?

**Comment by vasujain275** [comment]
@varshith257 We are focusing on production api cloud instance mainly. We added a development docker file because you suggested it. Docker for production is a higher priority right now.

**Comment by varshith257** [comment]
@ vasujain275 I think we have to keep docker files in the root. It's making hardship of the `npm run install` error as you faced. IMO it's better to keep it in the root folder. It's working fine there and if needed we remove the Dockerfiles from the root once it uploaded to dockerhub make a direct pull from docker hub and docker-compose files for both dev and prod containers and they  could be in the docker folder

File: Dockerfile.prod
Comment (varshith257) [reply Dockerfile.prod]: added

File: docker/docker-compose.dev.yaml
Comment (varshith257) [reply docker/docker-compose.dev.yaml:20]: noted

**Comment by varshith257** [comment]
> @ vasujain275
> 
> 1. I think we have to keep docker files in the root. It's making hardship of the `npm run install` error as you faced. IMO it's better to keep it in the root folder.It's working fine there.
> 2. And we have to make use of alpine in Dockerfile.dev too. As it making a size of 1.4GB I think there is no difference in making multistage builds and in most of the cases in industry practices they used alpine in the final stage they didn't report any issues regarding it.
> 3. It also affecting the use of volumes that I intended.

@palisadoes  Moved  docker files again to the root directory for the above reasons stated

**Comment by varshith257** [comment]
> [!NOTE]
I have observed a thing that `src/server.js` is a stale file and in package.json we are still using the server.js to run the production build files :
```
    "start": "cross-env pm2-runtime start ./build/server.js --watch",
    "prod": "cross-env NODE_ENV=production pm2-runtime start ./build/server.js",
```
It has to be pointed to` index.js `

- When I started a close look when I executed `npm run start` pointing to `build/index.js` produced this error:  
```

C:\Users\91970\Desktop\talawa\talawa-api>npm run start

> talawa-api@1.0.0 start
> cross-env pm2-runtime start ./build/index.js --watch

2024-02-28T00:17:02: PM2 log: Launching in no daemon mode
2024-02-28T00:17:02: PM2 log: [Watch] Start watching index
2024-02-28T00:17:02: PM2 log: App [index:0] starting in -fork mode-
2024-02-28T00:17:02: PM2 log: App [index:0] online
Error [ERR_REQUIRE_ESM]: require() of ES Module C:\Users\91970\Desktop\talawa\talawa-api\node_modules\nanoid\index.js from C:\Users\91970\Desktop\talawa\talawa-api\build\libraries\requestTracing.js not supported.
Instead change the require of index.js in C:\Users\91970\Desktop\talawa\talawa-api\build\libraries\requestTracing.js to a dynamic import() which is available in all CommonJS modules.
    at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
    at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\libraries\requestTracing.js:18:18)
    at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
    at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\libraries\logger.js:9:26)
    at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
    at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\libraries\index.js:31:14)
    at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
    at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\services\OrganizationCache\cacheOrganizations.js:16:21)
    at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
    at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\resolvers\DirectChat\organization.js:14:30)
    at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
    at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\resolvers\DirectChat\index.js:6:24)
    at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
    at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\resolvers\index.js:11:22)
    at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
    at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\index.js:41:21)
    at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\pm2\lib\ProcessContainerFork.js:33:23) {
  code: 'ERR_REQUIRE_ESM'
}
```

On close investigation I have found that nanoid stopped compatible with `commonjs` and new versions are compatible with `ES Modules`.

> [!CAUTION]
Though , resolved it by referring solution from StackOverflow i.e..downgrade it. It again started the same with `graphql-upload ` package recursively it getting errors with some package imports while compiler translating to `commonjs`  through  tsconfig.build.json

**Comment by vasujain275** [comment]
> @ vasujain275
> 
>     1. I think we have to keep docker files in the root. It's making hardship of the `npm run install` error as you faced. IMO it's better to keep it in the root folder.It's working fine there.
> 
>     2. And we have to make use of alpine in Dockerfile.dev too. As it making a size of 1.4GB I think there is no difference in making multistage builds and in most of the cases in industry practices they used alpine in the final stage they didn't report any issues regarding it.
> 
>     3. It also affecting the use of volumes that I intended.

@varshith257 

1. Keeping docker files in the specific "docker" directory will make root a lot more cluttered as @palisadoes stated. Please move all the files related to docker directory only. Happy to help out with errors but use this approach.
2. I strongly disagree that alpine should be used in Dockerfile.dev too, correct me if I am wrong @kb-0311. But according to my research we will face many complexities with alpine in development environment mainly - build steps, apollo server, anything ts as it needs a build step.
3. Some refs on why not use node:alpine in devlopement  - [Link](https://stackoverflow.com/questions/53024267/docker-are-node-alpine-images-at-the-end-smaller-than-full-node-images)
image
image
4. What is affecting the use of volumes? Please clarify.

**Comment by varshith257** [comment]
IMO, we have a old ` ts-config-build.json` and most of the packages only supporting the ES modules compatible but our ts-config.build .json emmits `commonjs` and it doesn't;t supporting them 

> @palisadoes @kb-0311 @noman2002
> 
> Note
> 
> I have observed a thing that `src/server.js` is a stale file and in package.json we are still using the server.js to run the production build files :
> 
> ```
>     "start": "cross-env pm2-runtime start ./build/server.js --watch",
>     "prod": "cross-env NODE_ENV=production pm2-runtime start ./build/server.js",
> ```
> 
> It has to be pointed to`index.js`
> 
> * When I started a close look when I executed `npm run start` pointing to `build/index.js` produced this error:
> 
> ```
> 
> C:\Users\91970\Desktop\talawa\talawa-api>npm run start
> 
> > talawa-api@1.0.0 start
> > cross-env pm2-runtime start ./build/index.js --watch
> 
> 2024-02-28T00:17:02: PM2 log: Launching in no daemon mode
> 2024-02-28T00:17:02: PM2 log: [Watch] Start watching index
> 2024-02-28T00:17:02: PM2 log: App [index:0] starting in -fork mode-
> 2024-02-28T00:17:02: PM2 log: App [index:0] online
> Error [ERR_REQUIRE_ESM]: require() of ES Module C:\Users\91970\Desktop\talawa\talawa-api\node_modules\nanoid\index.js from C:\Users\91970\Desktop\talawa\talawa-api\build\libraries\requestTracing.js not supported.
> Instead change the require of index.js in C:\Users\91970\Desktop\talawa\talawa-api\build\libraries\requestTracing.js to a dynamic import() which is available in all CommonJS modules.
>     at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
>     at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\libraries\requestTracing.js:18:18)
>     at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
>     at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\libraries\logger.js:9:26)
>     at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
>     at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\libraries\index.js:31:14)
>     at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
>     at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\services\OrganizationCache\cacheOrganizations.js:16:21)
>     at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
>     at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\resolvers\DirectChat\organization.js:14:30)
>     at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
>     at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\resolvers\DirectChat\index.js:6:24)
>     at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
>     at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\resolvers\index.js:11:22)
>     at Hook._require.Module.require (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\require-in-the-middle\index.js:101:39)
>     at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\build\index.js:41:21)
>     at Object.<anonymous> (C:\Users\91970\Desktop\talawa\talawa-api\node_modules\pm2\lib\ProcessContainerFork.js:33:23) {
>   code: 'ERR_REQUIRE_ESM'
> }
> ```
> 
> On close investigation I have found that nanoid stopped compatible with `commonjs` and new versions are compatible with `ES Modules`.
> 
> Caution
> 
> Though , resolved it by referring solution from StackOverflow i.e..downgrade it. It again started the same with `graphql-upload ` package recursively it getting errors with some package imports while compiler translating to `commonjs` through tsconfig.build.json

@xoldyckk  Need your guidance here. I have referred official documentation regarding this issue: and also found many projects facing this issue:
 https://www.typescriptlang.org/docs/handbook/modules/reference.html#module-format-detection
Should I change the module to `node16`  or `nodenext` instead ` commonjs` in `tsconfig.build.json` ?

**Comment by varshith257** [comment]
@palisadoes @noman2002  @kb-0311 Let's use  `node:<version>slim` for final image .
> [!NOTE]
 The most ideal Node.js Docker image would be a slimmed-down version of the operating system, based on a modern Debian OS, with a stable and active Long Term Support version of Node.js.

This is what I researched and choosing the right docker image is a ongoing process both alpine and slim has lower image size and slim we can say as a slimmed version of node image.

@vasujain275 I have referred article you provided which is the 5 year ago issue and latest is there is ongoing support to alpine image but using it is an experimental and unofficial node image .

For conclusion ,

1.  As of now we are going to use node:slim as final base image  to mitigate risks of alpine and reduced docker image size .
2. We are going to keep dockerfiles in root and me and @vasujain275 tried in different ways and leading to issues when moved to separate folders. The best common practises of docker is to keep docker files in the root directory and it's working fine there.

https://hub.docker.com/_/node
https://semaphoreci.com/blog/reduce-docker-image-size#best-strategies-to-slim-docker-images
https://snyk.io/blog/choosing-the-best-node-js-docker-image/

**Comment by varshith257** [comment]
> @varshith257 if the project can compile to javascript under es modules then sure

@xoldyckk I have tried but it is leading to new errors. This PR is going out of scope to support the production build of talawa-api .

**Comment by palisadoes** [comment]
Please work with @noman2002 on this

**Comment by kb-0311** [comment]
@varshith257 @vasujain275 I agree, there have also been issues with managing a few npm packages in `node-alipne` as well. Considering this repo has had issues with packages even during native development it was a good call to get rid of the risk while building the image in dev. 

My question is why is node-alpine used in production @vasujain275 ? Could we also opt for node-slim in prod as well? What do you guys think?

I think this PR is ready for a review.

**Comment by varshith257** [comment]
> @varshith257 @vasujain275 I agree, there have also been issues with managing a few npm packages in `node-alipne` as well. Considering this repo has had issues with packages even during native development it was a good call to get rid of the risk while building the image in dev. 
> 
> My question is why is node-alpine used in production @vasujain275 ? Could we also opt for node-slim in prod as well? What do you guys think?
> 
> I think this PR is ready for a review. 

For most of Devdependencies, it require to get the all packages installed without any issue can have some C related build.packages  as we used node-slim but in production the production dependencies can be  installed smoothly through node:alpine as they are managed with this lightweight image and it also.conteibutes to small image than node-slim.

As of now, the  node: alpine has good community support and working fine for production environment.

**Comment by vasujain275** [comment]
@kb-0311 @varshith257 We can safely use node:alpine for production Dockerfile becuase it only serves production ready .js build files. So we can be sure it will not have any dependency error because they are just html css js files it is serving but that is not the case for dev Dockerfile

**Comment by varshith257** [comment]
@vasujain275 @kb-0311  Is it fine as of now ?

**Comment by vasujain275** [comment]
> @vasujain275 @kb-0311 Is it fine as of now what we discussed till now?

Is the production Dockerfile fully working or still have the esmodule related errors you mentioned?

**Comment by varshith257** [comment]
> > @vasujain275 @kb-0311 Is it fine as of now what we discussed till now?
> 
> Is the production Dockerfile fully working or still have the esmodule related errors you mentioned?

It's works fine because we have pointed it to stale file which doesn't do anything just starts server. Need to raise an issue and work towards it .
As @xoldyckk suggestion, I changed to esnext of module in tsconfig.build file but it has a lot of errors to solve and it also a quite experimental as to check different setting to emmit build files without any issues.
As of now you can use it for cloud testing it down result any errors but can't access server.

**Comment by vasujain275** [comment]
> > > @vasujain275 @kb-0311 Is it fine as of now what we discussed till now?
> > 
> > 
> > Is the production Dockerfile fully working or still have the esmodule related errors you mentioned?
> 
>     * It's works fine because we have pointed it to stale file which doesn't do anything just starts server. Need to raise an issue and work towards it .
> 
>     * As @xoldyckk suggestion, I changed to esnext of module in tsconfig.build file but it has a lot of errors to solve and it also a quite experimental as to check different setting to emmit build files without any issues.
> 
>     * As of now you can use it for cloud testing it down result any errors but can't access server.

@varshith257 Okay got it, So here is what we should do  - 
1.  @varshith257 Remove the production docker installation from docs. As we have seprate documentation for them.
2. I will fix the talawa-api Production and build with help of @kb-0311 @xoldyckk, so we will not be to raise another Issue for it as I am already assigned the issue (#1428 ) .
3. In the mean time you can work on `talawa-admin` docker support. The issue that has been assigned to you already.
4. This way we can both efficiently work on the docker support for talawa applications in parallel. Because currently I am stuck in the issue #1428 because I need updated docker files to be merged. If you open another issue regarding the prodcution dockerfiles it will further hamper development of this issue .

**Comment by kb-0311** [comment]
@varshith257 @vasujain275 ok, got it, thanks.

So will you be working on the deprecated ts build modules? The reason why it points to `server.js` instead of `index.js` is because last year,  Eshaan and I had worked on the Apollo server version upgrade which was a huge mess due to a lot of incompatible and deprecated packages in talawa-api, during the start of last year's GSoC. Due to this we moved the server start point to index.js but did not focus on the build process as the build process was completely dysfunctional then and would have led to an additional downtime for our timelines. 

Would you like to take a look at the build process in a separate issue? You should work with xoldyckk on this. Tag me if you also want additional help : )

**Comment by varshith257** [comment]
@kb-0311 Sure let me raise an issue  and I had done a little research about it and let me work on it.

**Comment by vasujain275** [comment]
> @varshith257 @vasujain275 ok, got it, thanks.
> 
> So will you be working on the deprecated ts build modules? The reason why it points to `server.js` instead of `index.js` is because last year, Eshaan and I had worked on the Apollo server version upgrade which was a huge mess due to a lot of incompatible and deprecated packages in talawa-api, during the start of last year's GSoC. Due to this we moved the server start point to index.js but did not focus on the build process as the build process was completely dysfunctional then and would have led to an additional downtime for our timelines.
> 
> Would you like to take a look at the build process in a separate issue? You should work with xoldyckk on this. Tag me if you also want additional help : )

Sure I can work on it and we can work under this issue only - #1428  @kb-0311

**Comment by vasujain275** [comment]
@kb-0311 @varshith257 

We gotta work in parrallel to be efficient as I am idle now because this build error is needed to be fixed for me to solve #1428. That's why I suggested that I work on this issue of build erorr in talawa api and you can work on talawa admin docker support which is assigned to you. Happy to work on talawa admin one if you are more intrested in this/

**Comment by varshith257** [comment]
> > > > @vasujain275 @kb-0311 Is it fine as of now what we discussed till now?
> > > 
> > > 
> > > Is the production Dockerfile fully working or still have the esmodule related errors you mentioned?
> > 
> > 
> > ```
> > * It's works fine because we have pointed it to stale file which doesn't do anything just starts server. Need to raise an issue and work towards it .
> > 
> > * As @xoldyckk suggestion, I changed to esnext of module in tsconfig.build file but it has a lot of errors to solve and it also a quite experimental as to check different setting to emmit build files without any issues.
> > 
> > * As of now you can use it for cloud testing it down result any errors but can't access server.
> > ```
> 
> @varshith257 Okay got it, So here is what we should do -
> 
> 1. @varshith257 Remove the production docker installation from docs. As we have seprate documentation for them.
> 2. I will fix the talawa-api Production and build with help of @kb-0311 @xoldyckk, so we will not be to raise another Issue for it as I am already assigned the issue ([Cloud Based API Instance for Developers #1428](https://github.com/PalisadoesFoundation/talawa-api/issues/1428) ) .
> 3. In the mean time you can work on `talawa-admin` docker support. The issue that has been assigned to you already.
> 4. This way we can both efficiently work on the docker support for talawa applications in parallel. Because currently I am stuck in the issue [Cloud Based API Instance for Developers #1428](https://github.com/PalisadoesFoundation/talawa-api/issues/1428) because I need updated docker files to be merged. If you open another issue regarding the prodcution dockerfiles it will further hamper development of this issue .

1. It doesn't affect any docker files it's just a small change in the package.json command of npm start and make it compatible without any errors supporting the production build and access to localhost

**Comment by vasujain275** [comment]
Okay @varshith257, you fix this error and I will work on deploying to cloud CI/CD once this pr is merged. The Production Dockerfile points to stale server right? I can test the pipeline with just a stale server till you get the dependency bug fixed.

**Comment by varshith257** [comment]
@noman2002 @SiddheshKukade @EshaanAgg @DMills27 Can you please review this PR?

**Comment by varshith257** [comment]
@noman2002 @EshaanAgg Can I get reviewed this PR?

**Comment by varshith257** [comment]
@palisadoes Can this PR get reviewed? and I have uploaded video of related changes to access talawa-api using docker files for both dev and prod environment's.
It also delaying the work of cloud-instance issue.

**Comment by Cioppolo14** [comment]
@varshith257 I am working on it.

**Comment by varshith257** [comment]
> @varshith257 I am working on it. 

@Cioppolo14 Any updates on this to review.

**Comment by palisadoes** [comment]
This is an update on the PR merging freeze:

1. In the next few hours we will be merging the `develop-userTypeFix` branch into the `develop` branch. 
1. The `develop-userTypeFix` branch was created to fix a long standing design flaw where Admins were Admins of all organizations, not specific ones.
1. The `userType` field has been removed from the User collection and it has been replaced by an `appUserProfileId` field.
    1. This field is `null` if the user isn’t registered to use the apps. This will help people to add users manually during the event checkin process, or if an Admin wants to manually add someone in the Admin dashboard.
    1. When not `null` the `AppUserProfileID` will reference a `AppUserProfile` collection with App related information such as the organizations for which a user may be an Admin.
    1. The updated schema can be found here https://github.com/PalisadoesFoundation/talawa-api/blob/develop-userTypeFix/schema.graphql
    1. This is the parent issue that we have been using to track progress:
        1. https://github.com/PalisadoesFoundation/talawa-api/issues/1965
1. This merge will cause some conflicts in your PR.

We decided to do this at the beginning of the weekend to give us all time to adjust PR code and create bug fixes that may arise.

Update your code at or after midnight GMT on the morning of March 23, 2024. (5:30am IST). 

If your PRs have already been approved, request a re-review after fixing the conflicts and refactoring to the new `AppUserProfileID` methodology.

**Comment by palisadoes** [comment]
## The PR merging freeze is lifted.

1. We are working on bug fixes that may arise. These should be few and are being tracked here:
    1. https://github.com/orgs/PalisadoesFoundation/projects/24
1. Please update your PRs and local repos. Fix any new merge conflicts that may have occurred.
    
Background:

1. The `develop-userTypeFix` branch was created to fix a long standing design flaw where Admins were Admins of all organizations, not specific ones.
1. The `userType` field has been removed from the User collection and it has been replaced by an `appUserProfileId` field.
    1. This field is `null` if the user isn’t registered to use the apps. This will help people to add users manually during the event checkin process, or if an Admin wants to manually add someone in the Admin dashboard.
    1. When not `null` the `AppUserProfileID` will reference a `AppUserProfile` collection with App related information such as the organizations for which a user may be an Admin.
    1. The updated schema can be found here https://github.com/PalisadoesFoundation/talawa-api/blob/develop-userTypeFix/schema.graphql
    1. This is the parent issue that we have been using to track progress:
        1. https://github.com/PalisadoesFoundation/talawa-api/issues/1965
1. If your PRs have already been approved, request a re-review after fixing the conflicts and refactoring to the new `AppUserProfileID` methodology.

**Comment by Cioppolo14** [comment]
I am looking at this and hope to have any final comments or approval by the end of day Thursday.

File: .dockerignore
Comment (palisadoes) [inline .dockerignore:2]: Add this directory and anything else that's releveant in the `.gitignore` file

`+ image`

**Review by palisadoes — CHANGES_REQUESTED** [review]
1. See comments.
2. Does the `setup.ts` file and its dependencies need to be updated too?

**Comment by varshith257** [comment]
> 1. See comments.
> 2. Does the `setup.ts` file and its dependencies need to be updated too?

I think no need of any changes in the setup.ts as of now it's fine with Docker Setup 
image

File: .dockerignore
Comment (palisadoes) [reply .dockerignore:2]: Please comment on this

**Review by palisadoes — CHANGES_REQUESTED** [review]
See comment

**Comment by varshith257** [comment]
> See comment
 I added the files/folders that had large size and result in image size increase. The remaining doesn't make big change in image size. But ok, I will add the files/folders that doesn't needed in docker build

**Comment by palisadoes** [comment]
OK I'll merge

**Review by palisadoes — APPROVED** [review]

**Pull request merged**
